dirty_bcache: track unsynced and total dirty bytes, system-wide

strib opened the pull request:
And use these counts to control when writes are allowed to proceed.

I did my best to explain what's going on in the new comment above `folderBlockOps`, but please let me know if you need more clarification!

Issue: KBFS-1105
Issue: KBFS-1106
Issue: KBFS-1107

strib commented:
Arg, test failure.  Fixing now, hold off on review.

strib commented:
Ok, should be good to go again.

strib commented:
This is ready to review yet again.  I've confirmed that performance of writing a 128 MB file is the same both with and without this PR.  And when I bump up the parallel puts to 100 (and increase the dirty buffer to 50 MB), the write takes ~40% less time to complete (which is consistent with the previous results in KBFS-1066).

akalin-keybase commented on libkbfs/folder_block_ops.go:
```diff
@@ -67,6 +67,57 @@ func (si *syncInfo) DeepCopy(codec Codec) (*syncInfo, error) {
+//     creates a new copy of those blocks, and the whole size of that block
+//     (not just the newly-dirtied bytes) count for the total.  However,
+//     when the write gets replayed, folderBlockOps first subtracts those bytes
+//     from the system-wide numbers, since they are about the be replayed.
```
the -> to

akalin-keybase commented on libkbfs/dirty_bcache.go:
```diff
@@ -98,25 +131,119 @@ func (d *DirtyBlockCacheStandard) IsDirty(
+func (d *DirtyBlockCacheStandard) RequestPermissionToDirty(
+	estimatedDirtyBytes int64) DirtyPermChan {
+	if estimatedDirtyBytes < 0 {
+		panic("Must request permission for a positive number of bytes.")
```
shouldn't it be `<= 0` then?

akalin-keybase commented on libkbfs/dirty_file.go:
```diff
@@ -126,26 +171,50 @@ func (df *dirtyFile) resetSyncingBlocksToDirty() {
 	}
 	state.sync = blockSynced
+	df.dirtyBcache.BlockSyncFinished(state.syncSize)
+	//state.syncSize = 0
```
? debug code?

akalin-keybase commented:
First pass, mostly nits, still digesting. I'll wait for @jzila to take a look, too...

strib commented:
Thanks @akalin-keybase, nits fixed.

jzila commented on libkbfs/config_local.go:
```diff
@@ -569,7 +570,7 @@ func (c *ConfigLocal) ResetCaches() {
 	// Limit the block cache to 10K entries or 512 MB of bytes
 	c.bcache = NewBlockCacheStandard(c, 10000, 512*1024*1024)
-	c.dirtyBcache = NewDirtyBlockCacheStandard()
+	c.dirtyBcache = NewDirtyBlockCacheStandard(5<<20, 10<<20)
```
Comment here please to explain these numbers.

jzila commented on libkbfs/dirty_bcache.go:
```diff
@@ -43,9 +78,9 @@ func (d *DirtyBlockCacheStandard) Get(ptr BlockPointer, branch BranchName) (
-		d.dirtyLock.Lock()
-		defer d.dirtyLock.Unlock()
-		return d.dirty[dirtyID]
+		d.lock.Lock()
```
can this be a `RWMutex` where this is an `RLock()`?

jzila commented on libkbfs/dirty_bcache.go:
```diff
@@ -98,25 +131,119 @@ func (d *DirtyBlockCacheStandard) IsDirty(
-	d.dirtyLock.Lock()
-	defer d.dirtyLock.Unlock()
-	_, isDirty = d.dirty[dirtyID]
+	d.lock.Lock()
```
This can also be an `RLock()` if the mutex is converted to a `RWMutex`

jzila commented on libkbfs/dirty_bcache.go:
```diff
@@ -98,25 +131,119 @@ func (d *DirtyBlockCacheStandard) IsDirty(
+			newReq = true
+		}
+
+		if currentReq.respChan != nil && d.acceptNewWrite(currentReq.bytes) {
```
you can move the `currentReq.respChan` condition into a parent `if`

jzila commented on libkbfs/dirty_bcache.go:
```diff
@@ -98,25 +131,119 @@ func (d *DirtyBlockCacheStandard) IsDirty(
+	return canAccept
+}
+
+func (d *DirtyBlockCacheStandard) processPermission() {
```
I'd like a comment explaining the logic of this loop, or more detailed comments in the code, since there's some state that carries across iterations and it's difficult to follow.

jzila replied:
I'm convinced that the logic is correct, but it just takes a long time to understand that that is the case.

jzila commented on libkbfs/dirty_file.go:
```diff
@@ -184,5 +253,60 @@ func (df *dirtyFile) finishSync() error {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	listeners := df.errListeners
+	df.errListeners = nil
```
You can unlock after this.

jzila commented on libkbfs/dirty_file.go:
```diff
@@ -117,6 +151,17 @@ func (df *dirtyFile) setBlockSyncing(ptr BlockPointer) error {
 	state.sync = blockSyncing
+	block, err := df.dirtyBcache.Get(ptr, df.path.Branch)
+	if err != nil {
+		panic(err)
```
Might want to comment why this should never happen.

jzila commented:
LGTM modulo comments.

strib commented:
Thanks @jzila.  I addressed all your suggestions, and I fix one final thing: `RequestPermissionToDirty` needs to take a context, because in theory the request channel could fill up and cause blocking, so we need to return an error if the context is canceled.

@akalin-keybase: do you want to take another pass at this before I merge it?

strib commented:
Yep, no problem! Thanks for looking.

akalin-keybase commented:
lgtm!
